src/xml_parser_main_metadata_together.c

```diff
@@ -0,0 +1,371 @@
+}
+
+//TODO(amatej): add a flag that enforces pkg are in the same order (If possible?)
+//TODO(amatej): there is quite some overlap with this and cr_load_xml_files,
+//              we could use this api and just wrap it in cr_loax_xml_files?
+int cr_xml_parse_main_metadata_together(const char *primary_path,
+                                        const char *filelists_path,
+                                        const char *other_path,
+                                        cr_XmlParserNewPkgCb newpkgcb,
+                                        void *newpkgcb_data,
+                                        cr_XmlParserPkgCb pkgcb,
+                                        void *pkgcb_data,
+                                        cr_XmlParserWarningCb warningcb,
+                                        void *warningcb_data,
+                                        GError **err)
+{
+    int ret = CRE_OK;
+    CR_FILE *primary_f;
+    CR_FILE *filelists_f;
+    CR_FILE *other_f;
+    GError *tmp_err = NULL;
+
+    cr_CbData cbdata;
+    cbdata.in_progress_pkgs = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    cbdata.newpkgcb = newpkgcb;
+    cbdata.newpkgcb_data = newpkgcb_data;
+    cbdata.pkgcb = pkgcb;
+    cbdata.pkgcb_data = pkgcb_data;
+
+    //TODO(amatej): make filelists or other optional
+    assert(primary_path);
+    assert(filelists_path);
+    assert(other_path);
+    assert(newpkgcb || pkgcb);
+    assert(!err || *err == NULL);
+
+    primary_f = cr_open(primary_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    if (tmp_err) {
+        int code = tmp_err->code;
+        g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", primary_path);
+        return code;
+    }
+    filelists_f = cr_open(filelists_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    if (tmp_err) {
+        int code = tmp_err->code;
+        g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", filelists_path);
+        return code;
+    }
+    other_f = cr_open(other_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    if (tmp_err) {
+        int code = tmp_err->code;
+        g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", other_path);
+        return code;
+    }
+
+    cr_ParserData *primary_pd;
+    //TODO(amatej): replace do_files 0 by check if we are parsing filelists..
+    //TODO(amatej): if newpkgcb return NULL we should skip the pkg...
+    primary_pd = new_primary_parser_data(newpkgcb_primary, &cbdata, primary_pkgcb, &cbdata, warningcb, warningcb_data, 0);
+    cr_ParserData *filelists_pd;
+    filelists_pd = new_filelists_parser_data(newpkgcb_general, &cbdata, filelists_pkgcb, &cbdata, warningcb, warningcb_data);
+    cr_ParserData *other_pd;
+    other_pd = new_other_parser_data(newpkgcb_general, &cbdata, other_pkgcb, &cbdata, warningcb, warningcb_data);
+
+    char primary_is_done = 0;
+    char filelists_is_done = 0;
+    char other_is_done = 0;
```

Comment:
`char` capacity is 127, `XML_BUFFER_SIZE` is `8192`, which means the average package element size in the buffer in any of the 3 files *must* be >64 bytes.  This is a safe assumption but if `XML_BUFFER_SIZE` ever increased in the future (by a factor of 3 or 4) then it might result in an overflow.  Maybe uint32 should be used just to be safe.

Reply:
I misunderstood what they were for, then